### PR TITLE
adding error wrapper to align error reporting with codes

### PR DESCRIPTION
**Pull request by akoserwal**
Add Error/ServiceError interface as wrapper using https://pkg.go.dev/github.com/pkg/errors?utm_source=godoc
All services implementation will return the *model.ServiceError

Example of how to use in the services layer

`return model.NewWithCause(errors.ErrorGeneral, err, "Failed to process request")`

or

`return model.BadRequest(model.ErrorBadRequestReason, err)`

Clients can use the error codes to display user friendly errors

Suggestion
The error.go should be a separate package in the root directory.

### CODE REVIEW

**Comment by DGuhr** [comment]
3 small comments before I forget over easter:

1) I like it!
2) Idiomatic DDD would not allow to add an external dependency package to the domain layer. I see 3 options: 
    - move to application layer and map
    - don't use it at all
    - make an exception and put it there because of...
3) the [Readme](https://github.com/pkg/errors) states, that this is now in maintenance mode and will be replaced by a better error handling, including what this package would give us, in GO2.

**Comment by akoserwal** [comment]
I would suggest moving it as an error pkg at the top layer. 
Once https://go.googlesource.com/proposal/+/master/design/go2draft.md is available, we can perform the migration.

**Comment by DGuhr** [comment]
> I would suggest moving it as an error pkg at the top layer. Once https://go.googlesource.com/proposal/+/master/design/go2draft.md is available, we can perform the migration.

1) what do you mean by "top layer"? can you elaborate?
2) To give other ppl a bit more context around GO2: a little recherche shows that [GO2 is dead](https://www.techtarget.com/searchsoftwarequality/news/252526948/Go-updates-to-tackle-pain-points-but-Golang-2-is-dead) - but that seems more a naming/versioning thing. Looking at the [meta issue for error related stuff](https://github.com/golang/go/issues/40432) there are many proposals but I could not find anything that's in the "push it in" phase, so do not expect this to happen shortly (more: in a few years, perhaps)

**Comment by akoserwal** [comment]
what do you mean by "top layer"? can you elaborate?
-> We are moving towards a flat structure and I am suggesting moving error.go in at the error package at top level folder structure. It will be reusable across.

File: domain/model/Errors.go [hunk 1]
```diff
@@ -1,6 +1,164 @@
...
+
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+)
+
+const (
+	// ErrorCodePrefix Prefix error code
+	ErrorCodePrefix = "CIAM-Authz"
+	// ErrorForbidden Forbidden occurs when a user is not allowed to access the service
+	ErrorForbidden ServiceErrorCode = 4
```
[the remark this answers is not in the document]
Comment (akoserwal) [reply domain/model/Errors.go:15]: Arbitrary

File: domain/model/Errors.go [hunk 2]
```diff
@@ -1,6 +1,164 @@
...
+	// If the reason is unspecified, use the default
+	if reason != "" {
+		err.Reason = fmt.Sprintf(reason, values...)
+	}
+
+	return err
+}
+
+// ErrorList type
+type ErrorList []error
```
[the remark this answers is not in the document]
Comment (akoserwal) [reply domain/model/Errors.go:131]: Error slices will make it hard to figure out due to chains or errors. `AddErrors` will provide a list of errors.

[the remark this answers is not in the document]
Comment (Rajagopalan-Ranganathan) [reply domain/model/Errors.go:131]: As far as i can recollect, these are needed for more better UX usually, imagine a JSON - that comes in a request and in the server side validation there are different errors, its better to use this to return a list of validation errors. Can be extended based on our UI/UX needs to have [idx of JSON field] [Error] in future.

Comment (Rajagopalan-Ranganathan) [inline domain/model/Errors.go:131]: can we move the Global variables to the top?

File: domain/model/Errors.go [hunk 3]
```diff
@@ -1,6 +1,164 @@
...
+
+	// ErrorGeneral general error
+	ErrorGeneral ServiceErrorCode = 9
+	// ErrorGeneralReason general error reason
+	ErrorGeneralReason string = "Unspecified error"
+
+	// ErrorUnauthenticated unauthenticated error code
+	ErrorUnauthenticated ServiceErrorCode = 15
+	// ErrorUnauthenticatedReason unauthenticated error reason
+	ErrorUnauthenticatedReason string = "Account authentication could not be verified"
```
Comment (Rajagopalan-Ranganathan) [inline domain/model/Errors.go:27]: Is the use of "account" or "subject" be more relevant for our AuthZ?

**Comment by Rajagopalan-Ranganathan** [comment]
@akoserwal - Do we need a set of methods to capture and standardize gRPC errors? The file is about HTTP part alone? Do you think those errors are enough?

**Comment by Rajagopalan-Ranganathan** [comment]
Also, would this be the most correct location to format our error messages (that will be logged) in a more splunk friendly way? ( meaning: a config value if enabled formats the error logs in a more splunk friendly way)

**Comment by DGuhr** [comment]
I think  we need a design doc or a spike implementation before going further here. The Design Doc / Spike should imo clarify:
- In my first review I was (and still am) under the impression that this was just an example, and a design doc or sth like a more mature PoC would follow. I mainly focussed on the new methods, like StackTrace etc. - and that functionality seemed pretty nice to me. 
- Looking again, e.g. the "why" part with the arbitrary internal error codes is unclear for me. That may be my lack of internal context, asking myself how would it help with operating, integration in platform systems etc, because it adds additional complexity?!
- Also, there's already a [mapping](https://github.com/grpc-ecosystem/grpc-gateway/blob/main/runtime/errors.go#L36-L77) from HTTP to gRPC status codes in gRPC gateway ([usage example](https://blog.logrocket.com/guide-to-grpc-gateway/)), can these approaches be easily integrated / combined? Not investigated enough I think.
- architecture-aligned placement: I think neither putting it in the domain package nor defining another toplevel package are the best Ideas. Both would somehow "pollute" the domain layer with an external dependency. Also the "natural" first guess for me would be the application package. I would like to put some more thought into this.
 
Long story short: I like in general how this library would improve our services error handling, but I would veto on pushing this PR in its current form through, and would suggest us to have a team session to align here.

**Comment by Rajagopalan-Ranganathan** [comment]
As agreed, We are closing this PR. And will work on this on the subsequent small tasks as need arises
